### PR TITLE
Implement CRUD in the legacy REST API

### PR DESCRIPTION
**Pull request by emdashcodes**
Implements CRUD & correct getters/setters to avoid deprecation notices in all legacy API versions.

This does not handle variations updates. Variations need to be done for both the legacy versions and the current version, once #12228 lands.

### CODE REVIEW

File: includes/abstracts/abstract-wc-product.php [hunk 1]
```diff
@@ -2071,9 +2071,7 @@ public function has_file( $download_id = '' ) {
 	 * @return array
 	 */
 	public function get_files() {
-
-		$downloadable_files = array_filter( isset( $this->downloadable_files ) ? (array) maybe_unserialize( $this->downloadable_files ) : array() );
-
+		$downloadable_files = array_filter( ! empty( $this->get_downloads() ) ? (array) maybe_unserialize( $this->get_downloads() ) : array() );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: Wrapping method calls in empty will fatal error on PHP 5.2 and (3)?

File: includes/api/legacy/v1/class-wc-api-products.php [hunk 2]
```diff
@@ -268,14 +266,14 @@ private function query_products( $args ) {
...
+		$post = get_post( $product->get_id() );
 		return array(
 			'title'              => $product->get_title(),
-			'id'                 => (int) $product->is_type( 'variation' ) ? $product->get_variation_id() : $product->id,
-			'created_at'         => $this->server->format_datetime( $product->get_post_data()->post_date_gmt ),
-			'updated_at'         => $this->server->format_datetime( $product->get_post_data()->post_modified_gmt ),
-			'type'               => $product->product_type,
-			'status'             => $product->get_post_data()->post_status,
+			'id'                 => (int) $product->is_type( 'variation' ) ? $product->get_variation_id() : $product->get_id(),
+			'created_at'         => $this->server->format_datetime( $post->post_date_gmt ),
```
Comment (mikejolley) [inline includes/api/legacy/v1/class-wc-api-products.php]: There are methods for the dates if needed.

File: includes/api/legacy/v1/class-wc-api-products.php [hunk 3]
```diff
@@ -296,45 +294,46 @@ private function get_product_data( $product ) {
...
-			'featured_src'       => wp_get_attachment_url( get_post_thumbnail_id( $product->is_type( 'variation' ) ? $product->variation_id : $product->id ) ),
+			'featured_src'       => wp_get_attachment_url( get_post_thumbnail_id( $product->is_type( 'variation' ) ? $product->variation_id : $product->get_id() ) ),
 			'attributes'         => $this->get_attributes( $product ),
 			'downloads'          => $this->get_downloads( $product ),
-			'download_limit'     => (int) $product->download_limit,
-			'download_expiry'    => (int) $product->download_expiry,
-			'download_type'      => $product->download_type,
-			'purchase_note'      => apply_filters( 'the_content', $product->purchase_note ),
-			'total_sales'        => metadata_exists( 'post', $product->id, 'total_sales' ) ? (int) get_post_meta( $product->id, 'total_sales', true ) : 0,
+			'download_limit'     => (int) $product->get_download_limit(),
```
Comment (mikejolley) [inline includes/api/legacy/v1/class-wc-api-products.php]: Are casts needed here?

File: includes/api/legacy/v1/class-wc-api-products.php [hunk 4]
```diff
@@ -296,45 +294,46 @@ private function get_product_data( $product ) {
...
+			'download_expiry'    => (int) $product->get_download_expiry(),
+			'download_type'      => 'standard',
+			'purchase_note'      => apply_filters( 'the_content', $product->get_purchase_note() ),
+			'total_sales'        => metadata_exists( 'post', $product->get_id(), 'total_sales' ) ? (int) get_post_meta( $product->get_id(), 'total_sales', true ) : 0,
```
Comment (mikejolley) [inline includes/api/legacy/v1/class-wc-api-products.php]: Use get_total_sales()

File: includes/api/legacy/v1/class-wc-api-products.php [hunk 5]
```diff
@@ -406,22 +405,22 @@ private function get_images( $product ) {
...
-				$attachment_ids[] = get_post_thumbnail_id( $product->id );
+				$attachment_ids[] = $product->get_thumbnail_id();
 			}
 		} else {
 
-			// add featured image
-			if ( has_post_thumbnail( $product->id ) ) {
-				$attachment_ids[] = get_post_thumbnail_id( $product->id );
+			// Add featured image
+			if ( ! empty( $product->get_thumbnail_id() ) ) {
```
Comment (mikejolley) [inline includes/api/legacy/v1/class-wc-api-products.php]: empty fatal error

File: includes/api/legacy/v2/class-wc-api-products.php [hunk 6]
```diff
@@ -676,14 +679,14 @@ private function query_products( $args ) {
...
+		$post             = get_post( $product->get_id() );
 		return array(
 			'title'              => $product->get_title(),
-			'id'                 => (int) $product->is_type( 'variation' ) ? $product->get_variation_id() : $product->id,
-			'created_at'         => $this->server->format_datetime( $product->get_post_data()->post_date_gmt ),
-			'updated_at'         => $this->server->format_datetime( $product->get_post_data()->post_modified_gmt ),
-			'type'               => $product->product_type,
-			'status'             => $product->get_post_data()->post_status,
+			'id'                 => (int) $product->is_type( 'variation' ) ? $product->get_variation_id() : $product->get_id(),
+			'created_at'         => $this->server->format_datetime( $post->post_date_gmt ),
```
Comment (mikejolley) [inline includes/api/legacy/v2/class-wc-api-products.php]: get_date_created()

File: includes/api/legacy/v2/class-wc-api-products.php [hunk 7]
```diff
@@ -929,143 +919,126 @@ protected function save_product_meta( $product_id, $data ) {
...
-						'position'     => isset( $attribute['position'] ) ? (string) absint( $attribute['position'] ) : '0',
-						'is_visible'   => ( isset( $attribute['visible'] ) && $attribute['visible'] ) ? 1 : 0,
-						'is_variation' => ( isset( $attribute['variation'] ) && $attribute['variation'] ) ? 1 : 0,
-						'is_taxonomy'  => $is_taxonomy,
-					);
+					// Custom attribute - Add attribute to array and set the values.
+					$attribute_object = new WC_Product_Attribute();
+					$attribute_object->set_name( $attribute['name'] );
+					$attribute_object->set_options( $values );
+					$attribute_object->set_position( isset( $attribute['position'] ) ? (string) absint( $attribute['position'] ) : '0' );
```
Comment (mikejolley) [inline includes/api/legacy/v2/class-wc-api-products.php]: Position is integer, not string

File: includes/api/legacy/v2/class-wc-api-products.php [hunk 8]
```diff
@@ -1680,16 +1649,15 @@ private function get_images( $product ) {
...
 				// Otherwise use the parent product featured image if set
-				$attachment_ids[] = get_post_thumbnail_id( $product->id );
+				$attachment_ids[] = $product->get_thumbnail_id();
 			}
 		} else {
 
 			// Add featured image
-			if ( has_post_thumbnail( $product->id ) ) {
-				$attachment_ids[] = get_post_thumbnail_id( $product->id );
+			if ( ! empty( $product->get_thumbnail_id() ) ) {
```
Comment (mikejolley) [inline includes/api/legacy/v2/class-wc-api-products.php]: empty fatal

File: includes/api/legacy/v3/class-wc-api-products.php [hunk 9]
```diff
@@ -197,17 +198,17 @@ public function get_product( $id, $fields = null ) {
...
 		}
 
 		// Add grouped products data
 		if ( $product->is_type( 'grouped' ) && $product->has_child() ) {
 			$product_data['grouped_products'] = $this->get_grouped_products_data( $product );
 		}
 
-		if ( $product->is_type( 'simple' ) && ! empty( $product->post->post_parent ) ) {
-			$_product = wc_get_product( $product->post->post_parent );
+		if ( $product->is_type( 'simple' ) && ! empty( $product->get_parent_id() ) ) {
```
Comment (mikejolley) [inline includes/api/legacy/v3/class-wc-api-products.php]: empty

File: includes/api/legacy/v3/class-wc-api-products.php [hunk 10]
```diff
@@ -297,39 +298,43 @@ public function create_product( $data ) {
...
 			}
 
 			// Attempts to create the new product.
-			$id = wp_insert_post( $new_product, true );
+			$product->create();
+			$id = $product->get_id();
 
 			// Checks for an error in the product creation.
-			if ( is_wp_error( $id ) ) {
+			if ( ! $id > 0 ) {
```
Comment (mikejolley) [inline includes/api/legacy/v3/class-wc-api-products.php]: Little hard to read that conditonal. `$id <= 0`? Or yoda.

File: includes/api/legacy/v3/class-wc-api-products.php [hunk 11]
```diff
@@ -1655,92 +1621,87 @@ protected function save_product_meta( $product_id, $data ) {
...
-			// Download type.
-			if ( isset( $data['download_type'] ) ) {
-				update_post_meta( $product_id, '_download_type', wc_clean( $data['download_type'] ) );
+				$product->set_download_expiry( $data['download_expiry'] );
 			}
 		}
 
 		// Product url.
-		if ( 'external' === $product_type ) {
+		if ( 'external' === $product->get_type() ) {
```
Comment (mikejolley) [inline includes/api/legacy/v3/class-wc-api-products.php]: We have an is_type method which can be used instead.

**Review by mikejolley — COMMENTED** [review]
Notes added :)

**Comment by justinshreve** [comment]
Updates made & rebased.

**Comment by mikejolley** [comment]
Looks good here 👍

**Pull request merged**